### PR TITLE
Make nested IsInstantiables with Data in them work

### PR DESCRIPTION
**Pull request by mwachs5**
### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

   - bug fix         

#### API Impact

Fixes the bug, but the intended API does not change.

#### Backend Code Generation Impact

Stuff works that would have been impossible before to generate.

#### Desired Merge Strategy

   - Squash: The PR will be squashed and merged (choose this if you have no preference. 

#### Release Notes

Fix for #2760  -- Can now create non-RawModule `IsInstantiable` classes that contain @public reference to IOs and properly use them to connect to Instances.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: core/src/main/scala/chisel3/experimental/hierarchy/core/Lookupable.scala [hunk 1]
```diff
@@ -335,11 +335,21 @@ object Lookupable {
...
-          case Clone(x: ModuleClone[_]) => Some(x.ioMap)
-          case Proto(x: BaseModule) => Some(x.getChiselPorts.map { case (_, data) => data -> data }.toMap)
-          case _ => None
+
+        def getIoMap(hierarchy: Hierarchy[_]): Option[Map[Data, Data]] = {
+          hierarchy.underlying match {
+            case Clone(x: ModuleClone[_]) => Some(x.ioMap)
+            case Proto(x: BaseModule) => Some(x.getChiselPorts.map { case (_, data) => data -> data }.toMap)
+            case Clone(x: InstantiableClone[_]) => getIoMap(x._innerContext)
+            case Clone(x: InstanceClone[_]) => None // TODO: understand why this is OK
```
Comment (mwachs5) [inline core/src/main/scala/chisel3/experimental/hierarchy/core/Lookupable.scala:344]:
```suggestion
            case Clone(x: InstanceClone[_]) => None
```

File: core/src/main/scala/chisel3/Data.scala [hunk 2]
```diff
@@ -680,7 +680,7 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
           if (mod.flatMap(Builder.retrieveParent(_, Builder.currentModule.get)) == Builder.currentModule) =>
       case Some(_: UnconstrainedBinding) =>
       case _ =>
-        throwException(s"operand '$this' is not visible from the current module")
+        throwException(s"operand '$this' is not visible from the current module: (topBinding = ${topBindingOpt}")
```
Comment (mwachs5) [inline core/src/main/scala/chisel3/Data.scala:683]:
```suggestion
        throwException(s"operand '$this' is not visible from the current module: (topBinding = ${topBindingOpt})")
```

File: core/src/main/scala/chisel3/Data.scala [hunk 3]
```diff
@@ -680,7 +680,7 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
           if (mod.flatMap(Builder.retrieveParent(_, Builder.currentModule.get)) == Builder.currentModule) =>
       case Some(_: UnconstrainedBinding) =>
       case _ =>
-        throwException(s"operand '$this' is not visible from the current module")
+        throwException(s"operand '$this' is not visible from the current module: (topBinding = ${topBindingOpt})")
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/Data.scala:683]: While I grant that `blah is not visible` is not necessary the best user error message, I'm not sure that introducing the user to binding is quite right either. Is this a change you're intending to include?

**Review by jackkoenig — APPROVED** [review]
LGTM, some comments

File: core/src/main/scala/chisel3/Data.scala (hunk 3 above)
Comment (mwachs5) [reply core/src/main/scala/chisel3/Data.scala:683]: i mean, i found this useful in debugging the issue, but happy to take any other suggestions.

File: src/test/scala/chiselTests/experimental/hierarchy/InstanceSpec.scala [hunk 4]
```diff
@@ -403,6 +403,10 @@ class InstanceSpec extends ChiselFunSpec with Utils {
         MarkAnnotation("~Top|Top/i:HasMems>syncReadMem".rt, "SyncReadMem")
       )
     }
+    it("(3.p): should make connectable IOs on nested IsInstantiables that have IO Datas in them") {
+      val chirrtl = getFirrtlAndAnnos(new AddTwoNestedInstantiableData(4))._1.serialize
+      exactly(3, chirrtl.split('\n')) should include("i1.in <= i0.out")
```
[the remark this answers is not in the document]
Comment (mwachs5) [reply src/test/scala/chiselTests/experimental/hierarchy/InstanceSpec.scala:408]: me too, but i was matching a case above that used the pattern to get a `val chirrtl`.

File: core/src/main/scala/chisel3/experimental/hierarchy/core/Lookupable.scala [hunk 5]
```diff
@@ -335,11 +335,20 @@ object Lookupable {
...
-          case _ => None
+
+        def getIoMap(hierarchy: Hierarchy[_]): Option[Map[Data, Data]] = {
+          hierarchy.underlying match {
+            case Clone(x: ModuleClone[_]) => Some(x.ioMap)
+            case Proto(x: BaseModule) => Some(x.getChiselPorts.map { case (_, data) => data -> data }.toMap)
+            case Clone(x: InstantiableClone[_]) => getIoMap(x._innerContext)
+            case Clone(x: InstanceClone[_]) => None
+            case other => {
+              Builder.throwException(s"Internal Error! Unexpected case where we can't get IO Map: $other")
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/experimental/hierarchy/core/Lookupable.scala:346]:
```suggestion
              Builder.exception(s"Internal Error! Unexpected case where we can't get IO Map: $other")
```
whoops, got the method name wrong.

**Comment by mwachs5** [comment]
this fix may still be insufficient still trying to determine the corner case where this is so

File: src/test/scala/chiselTests/experimental/hierarchy/Examples.scala [hunk 6]
```diff
@@ -271,4 +271,43 @@ object Examples {
...
+    @public val leafOut = new LeafInstantiable(out)
+    @public val leafIn = new LeafInstantiable(in)
+    @public val nested = new NestedInstantiable(in = leafIn, out = leafOut)
+
+  }
+
+  class AddTwoNestedInstantiableData(width: Int) extends Module {
+    val in = IO(Input(UInt(width.W)))
+    val out = IO(Output(UInt(width.W)))
+    val addOneDef = Definition(new AddOneNestedInstantiableData(width))
```
Comment (mwachs5) [inline src/test/scala/chiselTests/experimental/hierarchy/Examples.scala:301]: this unit test is not testing the case where the Definition's parent is different than the Instance's parent. I think that case still has issues and will work on another unit test for that.

File: core/src/main/scala/chisel3/Data.scala (hunk 3 above)
Comment (mwachs5) [inline core/src/main/scala/chisel3/Data.scala:683]:
```suggestion
        throwException(s"operand '$this' is not visible from the current module")
```

reverting this change for now

File: src/test/scala/chiselTests/experimental/hierarchy/Examples.scala (hunk 6 above)
Comment (mwachs5) [reply src/test/scala/chiselTests/experimental/hierarchy/Examples.scala:301]: added the test case and it passes (my failing scenario was user error, so this fix appears good now)

File: src/test/scala/chiselTests/experimental/hierarchy/InstanceSpec.scala [hunk 7]
```diff
@@ -403,6 +403,16 @@ class InstanceSpec extends ChiselFunSpec with Utils {
...
     }
+    it("(3.p): should make connectable IOs on nested IsInstantiables that have IO Datas in them") {
+      val (chirrtl, _) = getFirrtlAndAnnos(new AddTwoNestedInstantiableData(4))
+      exactly(3, chirrtl.serialize.split('\n')) should include("i1.in <= i0.out")
+    }
+    it(
+      "(3.q): should make connectable IOs on nested IsInstantiables's Data when the Instance and Definition do not have the same parent"
+    ) {
+      val chirrtl = getFirrtlAndAnnos(new AddTwoNestedInstantiableDataWrapper(4))._1.serialize
+      exactly(3, chirrtl.split('\n')) should include("i1.in <= i0.out")
```
Comment (mwachs5) [inline src/test/scala/chiselTests/experimental/hierarchy/InstanceSpec.scala:414]:
```suggestion
      val (chirrtl, _) = getFirrtlAndAnnos(new AddTwoNestedInstantiableDataWrapper(4))
      exactly(3, chirrtl.serialize.split('\n')) should include("i1.in <= i0.out")
```
